src/core/classes/component-registry.ts

```diff
@@ -0,0 +1,171 @@
+import { allSettled } from '../../util/async';
+import type { ComponentProto } from '../../types';
+
+export interface ComponentRegistryOptions {
+	/** Path to the components */
+	path: string;
+	preload?: string[];
+}
+
+export interface ComponentProtoBase<Id extends string = string> {
+	id: Id;
+	require?: ComponentProto | readonly ComponentProto[];
+	optional?: string | readonly string[];
+}
+
+export interface AddEventPayload<T extends ComponentProto> {
+	id: string;
+	type?: string;
+	component: T;
+}
+
+export default class ComponentRegistry<T extends ComponentProto> extends EventTarget {
+	static type: string = 'unknown';
+
+	/** All imported components */
+	cache: Record<string, T> = {};
+
+	/** All components that are currently being loaded */
+	loading: Record<string, Promise<T>> = {};
+
+	/**
+	 * Same data as in loading, but as an array, used for aggregate promises.
+	 * IMPORTANT: Do NOT overwrite this array, only modify its contents.
+	 */
+	private loadingList: Promise<T>[] = [];
+
+	ready: Promise<T[]>;
```

Comment:
Shouldn't `ready` be a getter to become reactive? We change dynamically `loadingList` (`ready` depends on it). However, we set the value of `ready` once when creating an instance of a registry and never update it. As a result, we have `ready` that is always fulfilled, even though there might be promises that are not fulfilled in the `loadingList` array.
Or should we update `ready` every time we update `loadingList`? Or is there another option I'm not seeing?